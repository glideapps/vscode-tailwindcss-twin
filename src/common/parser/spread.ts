import * as nodes from "./nodes"
import { parse } from "./parse_regexp"
import { getVariant } from "./util"

interface VariantToken extends nodes.TokenString {
	type: nodes.NodeType.SimpleVariant | nodes.NodeType.ArbitraryVariant
}

export type SpreadDescription = {
	target: nodes.Classname | nodes.CssDeclaration | nodes.ArbitraryClassname
	value: string
	variants: VariantToken[]
	important: boolean
}

export function spread({ text, start = 0, end = text.length }: { text: string; start?: number; end?: number }) {
	interface Context {
		variants: VariantToken[]
		important: boolean
	}

	const items: SpreadDescription[] = []
	const emptyGroup: nodes.Group[] = []
	const emptyVariants: nodes.VariantSpan[] = []
	const notClosed: nodes.BracketNode[] = []

	const walk = (node: nodes.TwExpression | undefined, ctx: Context): void => {
		if (node == undefined) return

		if (nodes.NodeType.VariantSpan === node.type) {
			if (nodes.NodeType.ArbitraryVariant === node.variant.type && !node.variant.closed) {
				notClosed.push(node.variant)
				return
			}
			if (node.child == undefined) {
				emptyVariants.push(node)
				return
			}
			const variants = ctx.variants.slice()
			variants.push({ type: node.variant.type, ...getVariant(node.variant) })
			walk(node.child, { ...ctx, variants })
		} else if (nodes.NodeType.Group === node.type) {
			if (!node.closed) {
				notClosed.push(node)
				return
			}

			if (node.expressions.length === 0) {
				emptyGroup.push(node)
				return
			}

			node.expressions.forEach(n => walk(n, { ...ctx, important: ctx.important || node.important }))
		} else if (nodes.NodeType.CssDeclaration === node.type) {
			if (!node.closed) {
				notClosed.push(node)
				return
			}

			items.push({
				target: node,
				value: text.slice(...node.range),
				...ctx,
				important: ctx.important || node.important,
			})
		} else if (nodes.NodeType.ArbitraryClassname === node.type) {
			if (!node.closed) {
				notClosed.push(node)
				return
			}

			if (node.e && node.e.type === nodes.NodeType.WithOpacity && node.e.closed === false) {
				notClosed.push(node.e)
				return
			}

			items.push({
				target: node,
				value: text.slice(...node.range),
				...ctx,
				important: ctx.important || node.important,
			})
		} else if (nodes.NodeType.ClassName === node.type) {
			items.push({
				target: node,
				value: node.value,
				...ctx,
				important: ctx.important || node.important,
			})
		}
	}

	const program = parse({ text, start, end })
	program.expressions.forEach(expr => walk(expr, { variants: [], important: false }))

	return { items, emptyGroup, emptyVariants, notClosed }
}
