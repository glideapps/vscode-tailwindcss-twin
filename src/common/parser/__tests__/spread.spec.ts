import { NodeType } from ".."
import { spread } from "../spread"

it("parser util spread", () => {
	const result = spread({
		text: `bg-gray-500 padding[ 3rem ]!
		hover:(border-4 border-blue-500)
		[
			.group:hover,
			// .group:focus]
		]:!(text-red-700)// bg-primary
		/* text-error ()
		*/()
		custom/30 text-yellow-600/80 bg-gray-500/[.3]
		bg-[url(http://example)]
		text-[rgba(3 3 3 / .8)]/[.2] sm:hover: ( q-test`,
	})

	expect(result.items.map(i => i.target.type)).toEqual([
		NodeType.ClassName,
		NodeType.CssDeclaration,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ClassName,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
		NodeType.ArbitraryClassname,
	])

	expect(result).toMatchSnapshot()
})
