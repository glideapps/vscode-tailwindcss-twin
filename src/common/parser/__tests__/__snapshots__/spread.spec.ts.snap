// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser util spread 1`] = `
Object {
  "emptyGroup": Array [
    Object {
      "closed": true,
      "expressions": Array [],
      "important": false,
      "range": Array [
        161,
        163,
      ],
      "type": "Group",
    },
  ],
  "emptyVariants": Array [
    Object {
      "range": Array [
        273,
        279,
      ],
      "type": "VariantSpan",
      "variant": Object {
        "id": Object {
          "range": Array [
            273,
            278,
          ],
          "type": "Identifier",
          "value": "hover",
        },
        "range": Array [
          273,
          279,
        ],
        "type": "SimpleVariant",
      },
    },
  ],
  "items": Array [
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          0,
          11,
        ],
        "type": "ClassName",
        "value": "bg-gray-500",
      },
      "value": "bg-gray-500",
      "variants": Array [],
    },
    Object {
      "important": true,
      "target": Object {
        "closed": true,
        "expr": Object {
          "range": Array [
            20,
            26,
          ],
          "type": "CssExpression",
          "value": " 3rem ",
        },
        "important": true,
        "prop": Object {
          "range": Array [
            12,
            19,
          ],
          "type": "Identifier",
          "value": "padding",
        },
        "range": Array [
          12,
          27,
        ],
        "type": "CssDeclaration",
      },
      "value": "padding[ 3rem ]",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          38,
          46,
        ],
        "type": "ClassName",
        "value": "border-4",
      },
      "value": "border-4",
      "variants": Array [
        Object {
          "range": Array [
            31,
            36,
          ],
          "type": "SimpleVariant",
          "value": "hover",
        },
      ],
    },
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          47,
          62,
        ],
        "type": "ClassName",
        "value": "border-blue-500",
      },
      "value": "border-blue-500",
      "variants": Array [
        Object {
          "range": Array [
            31,
            36,
          ],
          "type": "SimpleVariant",
          "value": "hover",
        },
      ],
    },
    Object {
      "important": true,
      "target": Object {
        "important": false,
        "range": Array [
          111,
          123,
        ],
        "type": "ClassName",
        "value": "text-red-700",
      },
      "value": "text-red-700",
      "variants": Array [
        Object {
          "range": Array [
            67,
            107,
          ],
          "type": "ArbitraryVariant",
          "value": "
			.group:hover,
			// .group:focus]
		",
        },
      ],
    },
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          166,
          175,
        ],
        "type": "ClassName",
        "value": "custom/30",
      },
      "value": "custom/30",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "important": false,
        "range": Array [
          176,
          194,
        ],
        "type": "ClassName",
        "value": "text-yellow-600/80",
      },
      "value": "text-yellow-600/80",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "closed": true,
        "e": Object {
          "closed": true,
          "opacity": Object {
            "range": Array [
              208,
              210,
            ],
            "type": "Identifier",
            "value": ".3",
          },
          "range": Array [
            211,
            211,
          ],
          "type": "WithOpacity",
        },
        "expr": undefined,
        "important": false,
        "prop": Object {
          "range": Array [
            195,
            207,
          ],
          "type": "Identifier",
          "value": "bg-gray-500/",
        },
        "range": Array [
          195,
          211,
        ],
        "type": "ArbitraryClassname",
      },
      "value": "bg-gray-500/[.3]",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "closed": true,
        "e": undefined,
        "expr": Object {
          "range": Array [
            218,
            237,
          ],
          "type": "CssExpression",
          "value": "url(http://example)",
        },
        "important": false,
        "prop": Object {
          "range": Array [
            214,
            217,
          ],
          "type": "Identifier",
          "value": "bg-",
        },
        "range": Array [
          214,
          238,
        ],
        "type": "ArbitraryClassname",
      },
      "value": "bg-[url(http://example)]",
      "variants": Array [],
    },
    Object {
      "important": false,
      "target": Object {
        "closed": true,
        "e": Object {
          "closed": true,
          "opacity": Object {
            "range": Array [
              266,
              268,
            ],
            "type": "Identifier",
            "value": ".2",
          },
          "range": Array [
            265,
            269,
          ],
          "type": "WithOpacity",
        },
        "expr": Object {
          "range": Array [
            247,
            263,
          ],
          "type": "CssExpression",
          "value": "rgba(3 3 3 / .8)",
        },
        "important": false,
        "prop": Object {
          "range": Array [
            241,
            246,
          ],
          "type": "Identifier",
          "value": "text-",
        },
        "range": Array [
          241,
          269,
        ],
        "type": "ArbitraryClassname",
      },
      "value": "text-[rgba(3 3 3 / .8)]/[.2]",
      "variants": Array [],
    },
  ],
  "notClosed": Array [
    Object {
      "closed": false,
      "expressions": Array [
        Object {
          "important": false,
          "range": Array [
            282,
            288,
          ],
          "type": "ClassName",
          "value": "q-test",
        },
      ],
      "important": false,
      "range": Array [
        280,
        288,
      ],
      "type": "Group",
    },
  ],
}
`;
